File: src/teams_ai_planner/planner.py
"""High-level orchestration for fetching Teams data and planning actions."""

from __future__ import annotations

from datetime import datetime
from typing import Iterable, List, Optional, Protocol

from rich.console import Console
from rich.table import Table

from .graph import GraphClient
from .models import ActionPlan, ProposedAction, TeamsMessage, ensure_messages


class AnalyzerProtocol(Protocol):
    def suggest_actions(
        self,
        messages: Iterable[TeamsMessage],
        focus: Optional[str] = None,
        time_horizon_hours: int = 8,
    ) -> List[ProposedAction]:
        ...


class DailyActionPlanner:
    """Generate and optionally execute AI-suggested Teams follow-ups."""

    def __init__(
        self,
        graph_client: GraphClient | None,
        analyzer: AnalyzerProtocol,
        *,
        time_horizon_hours: int = 8,
    ) -> None:
        self._graph = graph_client
        self._analyzer = analyzer
        self._time_horizon_hours = time_horizon_hours
        self._console = Console()

    def generate_plan(
        self,
        team_id: str,
        channel_id: str,
        *,
        focus: Optional[str] = None,
        top: int = 40,
        lookback_hours: int = 24,
        offline_messages: Optional[Iterable[dict]] = None,
    ) -> ActionPlan:
        """Fetch Teams data (or accept offline messages) and analyze it."""

        if offline_messages is not None:
            raw_messages = list(offline_messages)
        else:
            if not self._graph:
                raise RuntimeError(
                    "Graph client is not configured. Provide --team-id/--channel-id or use offline JSON."
                )
            raw_messages = self._graph.get_recent_messages(
                team_id, channel_id, top=top, lookback_hours=lookback_hours
            )
        normalized: List[TeamsMessage] = ensure_messages(raw_messages)
        actions = self._analyzer.suggest_actions(
            normalized,
            focus=focus,
            time_horizon_hours=self._time_horizon_hours,
        )
        return ActionPlan(
            generated_at=datetime.utcnow(),
            timespan_hours=self._time_horizon_hours,
            message_sample_size=len(normalized),
            actions=actions,
        )

    def pretty_print(self, plan: ActionPlan) -> None:
        table = Table(title="Suggested follow-ups")
        table.add_column("Urgency", style="yellow")
        table.add_column("Title", style="bold")
        table.add_column("Details")
        table.add_column("Recipient")
        for action in plan.actions:
            table.add_row(
                action.urgency.capitalize(),
                action.title,
                action.details,
                action.recommended_recipient or "—",
            )
        self._console.print(table)

    def execute_followups(
        self,
        plan: ActionPlan,
        team_id: str,
        channel_id: str,
        *,
        dry_run: bool = True,
    ) -> list[dict]:
        """Send suggested follow-up messages back to the Teams channel."""

        if not self._graph:
            raise RuntimeError("Cannot send follow-ups without a Graph client.")

        responses = []
        for action in plan.actions:
            if not action.suggested_message:
                continue
            message = action.suggested_message
            if dry_run:
                self._console.print(
                    f"[cyan]DRY RUN[/cyan] Would send: {message[:120]}..."
                )
                continue
            responses.append(
                self._graph.send_channel_message(team_id, channel_id, message)
            )
        return responses

File: src/planner.ts
import { Analyzer } from "./analyzers/types.js";
import { GraphClient } from "./graphClient.js";
import { ActionPlan, ProposedAction, TeamsMessage, normalizeMessages } from "./types.js";

export interface PlanOptions {
  focus?: string;
  top?: number;
  lookbackHours?: number;
  offlineMessages?: Array<Record<string, any>>;
}

export interface DailyActionPlannerOptions {
  graphClient?: GraphClient;
  analyzer: Analyzer;
  timeHorizonHours: number;
}

export class DailyActionPlanner {
  private graphClient?: GraphClient;
  private analyzer: Analyzer;
  private timeHorizonHours: number;

  constructor(opts: DailyActionPlannerOptions) {
    this.graphClient = opts.graphClient;
    this.analyzer = opts.analyzer;
    this.timeHorizonHours = opts.timeHorizonHours;
  }

  async generatePlan(teamId: string, channelId: string, options: PlanOptions = {}): Promise<ActionPlan> {
    const { focus, top = 40, lookbackHours = 24, offlineMessages } = options;
    let rawMessages: Array<Record<string, any>>;
    if (offlineMessages) {
      rawMessages = [...offlineMessages];
    } else {
      if (!this.graphClient) {
        throw new Error("Graph client is not configured. Provide --team-id/--channel-id or use offline JSON.");
      }
      rawMessages = await this.graphClient.getRecentMessages(teamId, channelId, top, lookbackHours);
    }

    const normalized = normalizeMessages(rawMessages);
    const actions = await this.analyzer.suggestActions({
      messages: normalized,
      focus,
      timeHorizonHours: this.timeHorizonHours,
    });

    return {
      generatedAt: new Date().toISOString(),
      timespanHours: this.timeHorizonHours,
      messageSampleSize: normalized.length,
      actions,
    } satisfies ActionPlan;
  }

  printPlan(plan: ActionPlan, formatter: (action: ProposedAction) => void): void {
    plan.actions.forEach(formatter);
  }

  async executeFollowUps(plan: ActionPlan, teamId: string, channelId: string, { dryRun = true } = {}): Promise<any[]> {
    if (!this.graphClient) {
      throw new Error("Cannot send follow-ups without a Graph client.");
    }
    const responses: any[] = [];
    for (const action of plan.actions) {
      if (!action.suggestedMessage) continue;
      if (dryRun) {
        console.log(`[DRY RUN] Would send: ${action.suggestedMessage.slice(0, 120)}...`);
        continue;
      }
      const res = await this.graphClient.sendChannelMessage(teamId, channelId, action.suggestedMessage);
      responses.push(res);
    }
    return responses;
  }
}
